scripts/nmt/loss.py

```diff
@@ -43,9 +45,99 @@ def hybrid_forward(self, F, pred, label, valid_length):
+        self._axis = axis
+
+    def forward(self, is_train, req, in_data, out_data, aux):
+        inputs = in_data[0]
+        outputs = ((1 - self._epsilon) * inputs) + (self._epsilon / float(inputs.shape[self._axis]))
+        self.assign(out_data[0], req[0], outputs)
+
+    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
+        self.assign(in_grad[0], req[0], (1 - self._epsilon) * out_grad[0])
+
+
+@mx.operator.register('_smoothing_with_dim')
+class _SmoothingWithDimProp(mx.operator.CustomOpProp):
+    def __init__(self, epsilon=0.1, axis=-1):
+        super(_SmoothingWithDimProp, self).__init__(True)
+        self._epsilon = float(epsilon)
+        self._axis = int(axis)
+
+    def list_arguments(self):
+        return ['data']
+
+    def list_outputs(self):
+        return ['output']
+
+    def infer_shape(self, in_shape):
+        data_shape = in_shape[0]
+        output_shape = data_shape
+        return (data_shape,), (output_shape,), ()
+
+    def declare_backward_dependency(self, out_grad, in_data, out_data):
+        return out_grad
+
+    def create_operator(self, ctx, in_shapes, in_dtypes):
+        #  create and return the CustomOp class.
+        return _SmoothingWithDim(self._epsilon, self._axis)
+# pylint: enable=unused-argument
+
+class LabelSmoothing(HybridBlock):
+    """Applies label smoothing. See https://arxiv.org/abs/1512.00567.
+
+    Parameters
+    ----------
+    axis : int, default -1
+        The axis to smooth.
+    epsilon : float, default 0.1
+	The epsilon parameter in label smoothing
+    sparse_label : bool, default True
+        Whether input is an integer array instead of one hot array.
+    units : int or None
+	If units is given, it refers to the vocabulary size
+        Otherwise, it is infered from the input.
+    prefix : str, default 'rnn_'
+        Prefix for name of `Block`s
+        (and name of weight if params is `None`).
+    params : Parameter or None
+        Container for weight sharing between cells.
+        Created if `None`.
+    """
+    def __init__(self, axis=-1, epsilon=0.1, units=None,
+                 sparse_label=True, prefix=None, params=None):
+        super(LabelSmoothing, self).__init__(prefix=prefix, params=params)
+        self._axis = axis
+        self._epsilon = epsilon
+        self._sparse_label = sparse_label
+        self._units = units
+
+    def hybrid_forward(self, F, inputs, units=None):
+        if self._sparse_label:
+            assert units is not None or self._units is not None, \
+                'units needs to be given in function call or ' \
+                'instance initialization when sparse_label is False'
+            if units is None:
+                units = self._units
+            inputs = F.one_hot(inputs, depth=units)
+        if units is None and self._units is None:
+            return F.Custom(inputs, epsilon=self._epsilon, axis=self._axis,
+                            op_type='_smoothing_with_dim')
```

Comment:
We need to think how to avoid the CustomOp.

Reply:
Currently, it will not go to the CustomOp if the units is provided.